data/txHandler.go

```diff
@@ -759,3 +756,17 @@ func (handler *solicitedTxHandler) Handle(txgroup []transactions.SignedTxn) erro
 	}
 	return nil
 }
+
+func handleCapguard(c *util.ErlCapacityGuard, released, served bool) {
+	if c == nil {
+		return
+	}
+	if released {
+		if err := c.Release(); err != nil {
+			logging.Base().Warnf("Failed to release capacity to ElasticRateLimiter: %v", err)
+		}
+	}
+	if served {
+		c.Served()
+	}
+}
```

Comment:
There are 2 calls of `handleCapguard`:
```
			handleCapguard(wi.capguard, true, ac)
			if ac {
...
				continue
			}
...
			handleCapguard(wi.capguard, false, true)
```

So, if `ac == false`, `handleCapguard` is called twice:
true, false
false, true
resulting in: 
call 1
```
		if err := c.Release(); err != nil {
			logging.Base().Warnf("Failed to release capacity to ElasticRateLimiter: %v", err)
		}
```
call 2
```
	c.Served()
```

and if `ac == true`, `handleCapguard` is called once:
true, true
resulting in: 
```
		if err := c.Release(); err != nil {
			logging.Base().Warnf("Failed to release capacity to ElasticRateLimiter: %v", err)
		}
	c.Served()
```


```suggestion
func handleCapguard(c *util.ErlCapacityGuard) {
	if c == nil {
		return
	}
	if err := c.Release(); err != nil {
		logging.Base().Warnf("Failed to release capacity to ElasticRateLimiter: %v", err)
	}
	c.Served()
}
```

Reply:
The way it works now is intentional, and the way you're describing will not behave as needed.

A capacity guard, once vended, can do two things:

* "Released" -- gives capacity back to the rate limiter
* "Served" -- notifies the congestion controller (for calculation of service rate)

Both of these functions should be called once in total, with Release being the more critical of the two.

Release should be called *as soon as* the item has been dequeued. That is because the backlog now has capacity, and the capacity vending system should be able to hand out a new capacity guard.

Served should be called once the item is considered "fully serviced". In this case, it's once we know we are done working on it, or once we ship it off to the pool. That's because congestion control looks at the *service rate* to know how fast it can process items on the queue.

In your description of this behavior, you pointed out that in both situations (ac true or false), there is the right number of calls to call both Release and Served, exactly once.

The first call to `handleCapguard` -- if we find we have already committed the item, we can release the capacity and mark it served at the same time (`true, true`), because we just dequeued the item from the backlog (released) AND we discover that we've served the request (served).

The second call to `handleCapguard` -- If we had to do work (enqueue the work to the pool, etc), then we release the capacity asap, but we don't mark it served until this work item has moved on. That's why you'd see released first (`true, false`) and the served after (`false, true`)

In your implementation, Release and Serve are now directly linked -- if the capacity can be released, it is also served. This wouldn't cause any service errors necessarily, but it would incorrectly represent the service rate, as the actual processing time would never get factored into the service rate. The whole reason they're kept separate is to get an accurate service rate measurement.

Reply:
You are correct that there is currently not anything expensive happening between Release and Serve, and they are basically adjacent.

However, if we take this refactor, then they're *always* going to be consecutive, which defeats the whole reason they exist separately.

Imagine that between the `alreadyCommitted` check and the enqueue, we add an additional check that is very expensive.

```
dequeue from backlog
expensiveCheck(wi)
enqueue to pool
```

As currently written, the cost of the expensiveCheck gets factored into the calculated service rate. However, if we attempt to release and serve at the same time, we will either dismiss the expensive calculation from the service rate (not what we want), or we will return the capacity later than it could be returned (also not what we want).

In looking at this, I realize that actually the call `handleCapguard(true, ac)` should not be happening where it is, since the release of capacity can/should happen even before the `alreadyCommitted` check. Not sure why I put them in the same spot actually.

With that in mind, I actually think what we're looking for if you want to minimize the lines on txHandler, is two helpers in the erl package:

`erl.HandleCapguardRelease(capguard)` -- as soon as the wi is dequeued
`erl.HandleCapguardServed(capguard)` -- as soon as the work is "done"

The helpers can do the nil check and logging so they stay out of the way of txHandler. And this way, the separate concerns of Release and Serve are more clearly spelled out